File: app/csv_export.py
"""Small shared helper for streaming admin table data out as CSV.
Used by the Databases generic table browser and the Image Gallery admin
media/user exports — kept as one leaf module rather than duplicated
per-router since both just need "list of dicts -> CSV download"."""

import csv
import io
import re
from typing import Any

from fastapi.responses import StreamingResponse

_UNSAFE_FILENAME_RE = re.compile(r"[^A-Za-z0-9_.-]+")
# Column-name substrings that must never leave this process in a CSV, regardless of
# which table/query produced the rows — this is the last line of defense for any
# caller (including the generic cross-schema Databases browser), not just the
# call sites that remember to filter columns themselves.
_SENSITIVE_COLUMN_MARKERS = ("password", "secret", "token", "api_key")
# Excel/Sheets treat a leading =, +, -, or @ as the start of a formula. Prefixing
# with an apostrophe forces those clients to render the value as plain text.
_FORMULA_LEAD_CHARS = ("=", "+", "-", "@")


def _safe_filename(name: str) -> str:
    cleaned = _UNSAFE_FILENAME_RE.sub("_", str(name or "export")).strip("_") or "export"
    return cleaned[:120]


def _is_sensitive_column(key: str) -> bool:
    lowered = str(key or "").lower()
    return any(marker in lowered for marker in _SENSITIVE_COLUMN_MARKERS)


def _csv_cell(value: Any) -> str:
    text = "" if value is None else str(value)
    if text and text[0] in _FORMULA_LEAD_CHARS:
        return f"'{text}"
    return text


def rows_to_csv_response(rows: list[dict[str, Any]], filename: str) -> StreamingResponse:
    buffer = io.StringIO()
    fieldnames: list[str] = []
    for row in rows:
        for key in row.keys():
            if key not in fieldnames and not _is_sensitive_column(key):
                fieldnames.append(key)
    writer = csv.DictWriter(buffer, fieldnames=fieldnames, extrasaction="ignore")
    writer.writeheader()
    for row in rows:
        writer.writerow({key: _csv_cell(value) for key, value in row.items() if not _is_sensitive_column(key)})
    buffer.seek(0)
    safe_name = _safe_filename(filename)
    return StreamingResponse(
        iter([buffer.getvalue()]),
        media_type="text/csv",
        headers={"Content-Disposition": f'attachment; filename="{safe_name}.csv"'},
    )

File: lua/src/csv_export.lua
-- "List of dicts -> CSV download" helper. Port of app/csv_export.py,
-- including its two security-relevant rules: sensitive-column stripping
-- (password/secret/token/api_key substrings, regardless of which
-- table/query produced the rows) and CSV-formula-injection escaping
-- (Excel/Sheets treat a leading =+-@ as a formula).
local M = {}

local SENSITIVE_MARKERS = { "password", "secret", "token", "api_key" }

local function is_sensitive_column(key)
  local lowered = tostring(key or ""):lower()
  for _, marker in ipairs(SENSITIVE_MARKERS) do
    if lowered:find(marker, 1, true) then return true end
  end
  return false
end

local function csv_cell(value)
  local text = (value == nil) and "" or tostring(value)
  local lead = text:sub(1, 1)
  if text ~= "" and (lead == "=" or lead == "+" or lead == "-" or lead == "@") then
    text = "'" .. text
  end
  -- RFC4180 quoting: always quote and double-up embedded quotes; simplest
  -- correct behavior regardless of whether the cell needs it.
  return '"' .. text:gsub('"', '""') .. '"'
end

function M.rows_to_csv_text(rows)
  local fieldnames, seen = {}, {}
  for _, row in ipairs(rows or {}) do
    for key in pairs(row) do
      if not seen[key] and not is_sensitive_column(key) then
        seen[key] = true
        fieldnames[#fieldnames + 1] = key
      end
    end
  end
  table.sort(fieldnames) -- Lua's pairs() has no stable order; sorted is deterministic (Python's dict-insertion order isn't reproducible here anyway)
  local lines = {}
  local header = {}
  for _, k in ipairs(fieldnames) do header[#header + 1] = csv_cell(k) end
  lines[#lines + 1] = table.concat(header, ",")
  for _, row in ipairs(rows or {}) do
    local cells = {}
    for _, k in ipairs(fieldnames) do cells[#cells + 1] = csv_cell(row[k]) end
    lines[#lines + 1] = table.concat(cells, ",")
  end
  return table.concat(lines, "\r\n") .. "\r\n"
end

local UNSAFE_FILENAME = "[^%w_.-]+"
function M.safe_filename(name)
  local cleaned = tostring(name or "export"):gsub(UNSAFE_FILENAME, "_"):gsub("^_+", ""):gsub("_+$", "")
  if cleaned == "" then cleaned = "export" end
  return cleaned:sub(1, 120)
end

-- Returns (status, body_string, headers) ready to hand back from an
-- httpd.route handler (httpd.lua passes non-table bodies through as-is and
-- respects any Content-Type override in the extra-headers table).
function M.csv_response(rows, filename)
  local body = M.rows_to_csv_text(rows)
  local headers = {
    ["Content-Type"] = "text/csv",
    ["Content-Disposition"] = string.format('attachment; filename="%s.csv"', M.safe_filename(filename)),
  }
  return 200, body, headers
end

return M
